### PR TITLE
Support arbitrary state shapes using selectLocationState

### PR DESCRIPTION
**Pull request by iansinnott**
This PR is meant to continue the discussion of how to best handle a use case like Immutable.js, where the state shape of the app differs from what this lib expects. If you like what I've done here then by all means let's get it merged, but I'm also open to any and all feedback. Ultimately I really just want to be able to use this for my own projects which happen to use Immutable, but this change would also help anyone else dealing with a different state shape.

## What is this

This came out of the discussion in #20, but basically it would help widespread adoption of this lib if it was not dependent on using plain objects for your state shape.

## What changed

* **New options:** `selectLocationState` and `selectTitleState`. These are both functions that replace the use case of `location` and `title` options.

Previously the `location` option (referened as `locationKey` internally) allowed the user to specify _where_ in state the location reducer state would be held. However, this assumed that the user would be using plain javascipt objects as their state shape. By replacing this string option with a selector function we allow much more flexibility over how the location state is stored and accessed.

Before:

```js
const { reducer, middleware, enhancer } = connectRoutes(history, routesMap, {
  location: 'customLocationKey',
});
```
After:

```js
const { reducer, middleware, enhancer } = connectRoutes(history, routesMap, {
  selectLocationState: state => state.customLocationKey,
});
```

Or with another library like Immutable.js. In this example `state` is an `Immutable.Map`:

```js
const { reducer, middleware, enhancer } = connectRoutes(history, routesMap, {
  selectLocationState: state => state.get('customLocationKey'),
});
```

## Breaking Changes ⚠

Since `selectLocationState` makes the location option unecessary it has been removed. But since it was publicly exported as part of the API this represents a breaking change. In fact this did break `redux-first-router-link` during development because it depends on the `locationKey`. This is simple enough to solve by exporting the `selectLocationState` function instead of the string key, so there shouldn't be any long term issues here and I will PR the other repo too if you want to move forward with this change. But a breaking change like this would warrant a major version bump. Just something to be aware of.

### CODE REVIEW

**Comment by faceyspacey** [comment]
Excellent, I'll be getting back to you in depth when I get a moment. Thanks for this. 1st serious PR on my work by the way.

**Comment by iansinnott** [comment]
Cheers @faceyspacey 🍻 

For a live example, here is my own repo where I was working on this. I had `yarn link`ed the local changes, so building that repository won't work. But you can see how I am using the library seamlessly with Immutable: https://github.com/iansinnott/try-redux-first-router/blob/master/client/index.js#L64

@oleksiipysanko this might be of interest to you as well.

File: src/connectRoutes.js [hunk 1]
```diff
@@ -115,8 +115,8 @@ export default (
   }
 
   const {
-    location: locationKey = 'location',
-    title: titleKey = 'title',
+    selectLocationState = state => state.location,
```
Comment (faceyspacey) [inline src/connectRoutes.js]: let's keep the original external options names (`location`, `title`) so we don't need to introduce a breaking change. And let's still except strings. If a string is passed, create a selector function once right here that can be reused. A function named `selectLocation/Title` state like you already have can be used throughout the rest of the code.

Comment (faceyspacey) [inline src/connectRoutes.js]: Other than that, there's also the Flow errors which you can see in Travis. You can run `npm run flow` to see em from the console. You need to change the signature for `location` + `title` in `src/flowTypes.js`. Obviously now it's signature supports both a string and a function.

Comment (faceyspacey) [inline src/connectRoutes.js]: then update `docs/connectRoutes.md` to reflect these changes to the option, and we should be good to go.

Comment (iansinnott) [inline src/connectRoutes.js]: Ok, to be clear, were you thinking to overload the `title` and `location` options to allow functions and strings or to keep them around in addition to `selectLocatin/TitleState` ?

Comment (faceyspacey) [inline src/connectRoutes.js]: Exactly

Comment (iansinnott) [inline src/connectRoutes.js]: OK, so the first of the two options then?

I think flow might complain about the overloading of `string | Function`. We'll see

File: src/connectRoutes.js [hunk 2]
```diff
@@ -317,6 +329,9 @@ export default (
     }
 
     if (typeof window !== 'undefined') {
+      // `kind` is typed as ?string which will not throw if passed to .test.
+      // Typing kind as string threw other flow errrs in the codebase
+      // $FlowIssue
```
Comment (iansinnott) [inline src/connectRoutes.js]: Open to suggestions as to how to improve type coverage here. Since kind is `?string` flow does no like passing it to `test`. However going from `?string` -> `string` caused a slew of other flow related issues in the codebase.

File: src/connectRoutes.js [hunk 3]
```diff
@@ -115,7 +115,11 @@ export default (
   }
 
   const {
+    // Flow doesn't like defaulting to string or func here, although it does not
+    // complain if we _do not_ provide a default value
+    // $FlowIssue
```
Comment (iansinnott) [inline src/connectRoutes.js]: This is not pretty, but flow was not having a good time with default arguments here. However, flow seems to know the type but type this incorrectly. Maybe I'm missing something, but flow seemed to be saying two conflicting things at once:

```
src/connectRoutes.js:118
118:     location: locationKey = 'location',
                   ^^^^^^^^^^^ function type. This type is incompatible with
118:     location: locationKey = 'location',
                   ^^^^^^^^^^^ string

src/connectRoutes.js:118
118:     location: locationKey = 'location',
                   ^^^^^^^^^^^ string. This type is incompatible with
118:     location: locationKey = 'location',
                   ^^^^^^^^^^^ function type
```

File: src/connectRoutes.js [hunk 4]
```diff
@@ -126,6 +130,14 @@ export default (
...
   }: Options = options
 
+  const selectLocationState = typeof locationKey === 'function'
+    ? locationKey
+    : state => state[locationKey]
+
+  const selectTitleState = typeof titleKey === 'function'
+    ? titleKey
+    : state => state[titleKey]
+
```
Comment (iansinnott) [inline src/connectRoutes.js]: This is where the overloading happens. I'm not a huge fan of overloading like this (overloading in general reminds me of jQuery... 😱) but I understand the desire not to change the API yet.

File: src/connectRoutes.js (hunk 1 above)
Comment (faceyspacey) [inline src/connectRoutes.js]: oh my bad, not in addition. just the original 2 like u did. everything is looking perfect, including the docs.

...as long as u do `if typeof === 'function'` like u did flow will be able to infer.

...i saw the comment about the `kind`. that must be a new flow issue. each update to flow, it gets better and can find more things. Do you mind just checking the existence of the `kind` key rather than using `$FlowIssue`.

File: src/connectRoutes.js (hunk 4 above)
Comment (faceyspacey) [inline src/connectRoutes.js]: that too, but u often see the pattern of being able to provide a string or a selector. it's not uncommon even in 2017.

File: src/connectRoutes.js (hunk 2 above)
Comment (faceyspacey) [inline src/connectRoutes.js]: do this:

```
if (typeof window !== 'undefined' && kind)
```

Comment (faceyspacey) [inline src/connectRoutes.js]: fyi, later i'll check if we even need the `?`. it should always be there. either way, if it's not, we wont past that test anyway.

File: src/connectRoutes.js (hunk 3 above)
Comment (faceyspacey) [inline src/connectRoutes.js]: don't default there. default manually as part of a typeof function | string | *undefined* check (all in one place)

Comment (faceyspacey) [inline src/connectRoutes.js]: but im surprised that that they don't let u set a default value of one of the expected types. weird.

File: src/connectRoutes.js (hunk 4 above)
Comment (faceyspacey) [inline src/connectRoutes.js]: mainly by the way i can foresee a potential future where the number of options gets out of hand, and im trying to avoid that. less API surface area to think about.

File: src/connectRoutes.js (hunk 2 above)
Comment (iansinnott) [inline src/connectRoutes.js]: Yeah, I was thinking it should probably always be defined but didn't want to change something like that since it seems unrelated to this PR

**Comment by faceyspacey** [comment]
can u give me access to ur fork so i can push a test to it

**Comment by faceyspacey** [comment]
here's the test by the way. u could just add it to __tests__connectRoutes.js at the very bottom:


```js
it('title and location options as selector functions', () => {
  const { middleware, reducer, enhancer } = setup('/first', {
    title: state => state.title,
    location: state => state.location
  })
  const middlewares = applyMiddleware(middleware)

  const rootReducer = (state = {}, action = {}) => ({
    location: reducer(state.location, action),
    title: action.type
  })

  const store = createStore(rootReducer, middlewares, enhancer)
  const action = store.dispatch({ type: 'FIRST' }) /*? $.meta */

  store.getState() /*? $.location */

  expect(action).toMatchSnapshot()
})
```

...its good to know how u would go about writing the test. ...basically in `setup()` we do all the work related to calling `connectRoutes`.

**Comment by iansinnott** [comment]
Awesome! Thanks for writing a test. I was looking the text file and thinking hm 🤔 , I bet @faceyspacey could whip up a useful test much faster than me. But yeah adding tests is certainly best practice so I'm happy to see how you're doing it.

**Comment by iansinnott** [comment]
Adding tests and pushing...

**Comment by faceyspacey** [comment]
nicely done my friend!

**Comment by iansinnott** [comment]
Happy to contribute. Your blog post + what the formidable guys wrote really resonated. This is the router now.

Thanks for being super responsive on this. Always makes the PRs so much more fun 👏 🍻

**Comment by faceyspacey** [comment]
definitely!

**Comment by faceyspacey** [comment]
added ur quote at the top of today's article by the way (hope that's ok): https://medium.com/faceyspacey/redux-first-router-lookin-sexy-on-code-sandbox-d9d9bea15053

**Comment by iansinnott** [comment]
Yeah, of course! Great to see Berkeley at the top of that list too. He's the one who introduced me to this
